### ISSUE TITLE
FulfilledPromise not invoking callbacks immediately

### ISSUE THREAD

**Issue by mikegreiling** [guzzle/promises#14]
The comments for FulfilledPromise indicate that any calls to `then()` will invoke the callback immediately, but it seems to just add it to `queue()`.

Is the documentation wrong or is this a bug?

**Comment by mtdowling** [comment]
This is not a bug. You need to either call wait() on a promise or manually tick the queue.

A queue is used to invoke callbacks to prevent recursion and to guarantee a predictable order for callbacks -- for example, if we invoked callbacks immediately when a promise is fulfilled without queueing them, then we could invoke callbacks out of the order in which they were added to the promise (e.g., callbacks are added on a pending promise, pending promise is fulfilled, callbacks are added, then the queue is ticked to invoke callbacks).

What part of the documentation is wrong?

**Comment by mikegreiling** [comment]
I guess I misunderstood the use of the word "immediately" in the following comment blocks and documentation sections.  Perhaps different wording would be appropriate.

https://github.com/guzzle/promises/blob/master/src/FulfilledPromise.php#L7
https://github.com/guzzle/promises/blob/master/src/RejectedPromise.php#L7
https://github.com/guzzle/promises/blob/master/src/Promise.php#L42
https://github.com/guzzle/promises/blob/master/README.md#fulfilledpromise

**Comment by mtdowling** [comment]
Yeah, the wording there is a bit ambiguous as to what "immediately" means. The intent is that there is no fulfillment required for the promise to resolve the callbacks. Maybe it should be changed to "immediately on the next tick"

**Comment by Zxurian** [comment]
so this was actually referenced in a problem I'm trying to solve with https://github.com/aws/aws-sdk-php .
I'm trying to write a routine such that there is a main loop (Loop-A) that runs over a fixed list checking for certain triggers. If a trigger is found, it initiates an asynchronous upload that returns a Guzzle Promise, which I attach a `->then()` to. This issue came up because the callbacks I'm attaching via `then()` aren't firing upon completion of the upload because of the above problem of not resolving until the next "tick' of the event loop.

Is there anything I can run as part of the main loop (loop-a) that will check through all existing promises to see if any are fulfilled and then resolve them? Or possibly a better solution, a way of having the promises being fulfilled without needing an additional tick of the queue of the callbacks are registered to the specific promises?

**Comment by mtdowling** [comment]
@Zxurian can you provide a reproducible code example I can see? My first thoughts are that you need to chain your promises if you want your callbacks to be invoked without an extra tick. You can get access to the TaskQueue and invoke `run()` in your loop as well. See  https://github.com/guzzle/promises/blob/master/src/TaskQueue.php

**Comment by Zxurian** [comment]
@mtdowling using the following code example with https://github.com/aws/aws-sdk-php
```php
<?php

$uploading = 0;

$AWS = new Aws\Sdk([
    'version'    => 'latest',
    'region'    => 'us-east-1',
    'credentials'    => [
        'key'        => '***',
        'secret'    => '***',
    ]
]);

$S3 = $AWS->createS3();

echo "beginning first promise\n";
$uploading ++;
$promise1 = $S3->uploadAsync('tantor-books', 'Test/test.xml', fopen('/mnt/olympus/Books/0087_CoalHuman/Download/0087_CoalHuman.xml', 'r'));
$promise1->then(function() use ($uploading) {
    $uploading --;
});

echo "beginning second promise\n";
$uploading ++;
$promise2 = $S3->uploadAsync('tantor-books', 'Test/test.zip', fopen('/mnt/olympus/Books/0087_CoalHuman/Download/0087_CoalHuman.zip', 'r'));
$promise2->then(function() use ($uploading) {
    $uploading --;
});

echo "beginning loop\n";
while ($uploading != 0) {
    echo 'uploading: '.$uploading."\n";
}
```

TL;DR - The script will never exit as the `then()` functions never get called.
I'm also having a discussion at https://gitter.im/aws/aws-sdk-php about it and @jeskew says this is by design. and that I'll need to manually tick the queue forward before the uploads will actually start. What I'd _like_ is to start files uploading while the main loop is still going, then `wait()` on anything that's still transferring at the end before proceeding, and it's sounding like in order to do that, I need to manually tick the Guzzle queue after each call to `uploadAsync()` in order to start it.

**Comment by berarma** [comment]
@Zxurian, I found this while working on a similar problem. PHP isn't multi-threaded so while your loop is running nothing else will run. That includes your S3 client code, responsible of fulfilling the promises, and the promise queue. Async operations are simulated by waiting on external events and queueing tasks.

You could add this to your loop:
```php
GuzzleHttp\Promise\queue->run();
```

But it won't have anything to run because the client code didn't have a chance to fulfill the promises. You should be ticking your S3 client code instead (in case that's possible) or wait on it. When you call the tick or wait operations on the client it simulates multi-threading by chaining all the operations. So you should do this instead of creating your own wait loop, or you should be able to do reproduce the same behavior inside your loop.

I guess it should be done like this:
```php
<?php

$uploading = 0;

$AWS = new Aws\Sdk([
    'version'    => 'latest',
    'region'    => 'us-east-1',
    'credentials'    => [
        'key'        => '***',
        'secret'    => '***',
    ]
]);

$S3 = $AWS->createS3();

echo "beginning first promise\n";
$uploading ++;
$promise1 = $S3->uploadAsync('tantor-books', 'Test/test.xml', fopen('/mnt/olympus/Books/0087_CoalHuman/Download/0087_CoalHuman.xml', 'r'));
$promise1->then(function() use ($uploading) {
    $uploading --;
    echo 'uploading: '.$uploading."\n";
});

echo "beginning second promise\n";
$uploading ++;
$promise2 = $S3->uploadAsync('tantor-books', 'Test/test.zip', fopen('/mnt/olympus/Books/0087_CoalHuman/Download/0087_CoalHuman.zip', 'r'));
$promise2->then(function() use ($uploading) {
    $uploading --;
    echo 'uploading: '.$uploading."\n";
});

echo "beginning loop\n";
// Call here your S3 wait function or unwrap the promises that in turn will do the same.
GuzzleHttp\Promise\unwrap([$promise1, $promise2]);
```

Please, correct me if I'm wrong.

**Comment by GrahamCampbell** [comment]
🎉
